### ISSUE TITLE
ClientCacheQuorumWriteTest Failrues

### ISSUE THREAD

**Issue by hasancelik** [hazelcast/hazelcast#12309]
**ClientCacheQuorumWriteTest.put_quorum**
```
com.hazelcast.cache.CacheNotExistsException: Cache /hz/quorume149952d-f46d-4cfc-89ec-19ab09a63b4fWRITE is already destroyed or not created yet, on Member [127.0.0.1]:9203 - a173f188-f2a0-426a-b91a-916e0b3c25d0 this
	at com.hazelcast.client.impl.protocol.task.cache.AbstractCacheMessageTask.getOperationProvider(AbstractCacheMessageTask.java:51)
	at com.hazelcast.client.impl.protocol.task.cache.CachePutMessageTask.prepareOperation(CachePutMessageTask.java:46)
	at com.hazelcast.client.impl.protocol.task.AbstractPartitionMessageTask.processMessage(AbstractPartitionMessageTask.java:62)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.initializeAndProcessMessage(AbstractMessageTask.java:123)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.run(AbstractMessageTask.java:103)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:155)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:125)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:100)
	at ------ submitted from ------.(Unknown Source)
	at java.lang.Thread.getStackTrace(Thread.java:1117)
	at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:96)
	at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:33)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.get(AbstractInvocationFuture.java:183)
	at com.hazelcast.client.util.ClientDelegatingFuture.get(ClientDelegatingFuture.java:137)
	at com.hazelcast.client.util.ClientDelegatingFuture.get(ClientDelegatingFuture.java:128)
	at com.hazelcast.client.cache.impl.AbstractClientInternalCacheProxy.putSyncInternal(AbstractClientInternalCacheProxy.java:421)
	at com.hazelcast.client.cache.impl.AbstractClientCacheProxy.put(AbstractClientCacheProxy.java:261)
	at com.hazelcast.client.cache.impl.ClientCacheProxy.put(ClientCacheProxy.java:94)
	at com.hazelcast.client.cache.impl.ClientCacheProxy.put(ClientCacheProxy.java:191)
	at com.hazelcast.quorum.cache.CacheQuorumWriteTest.put_quorum(CacheQuorumWriteTest.java:83)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:95)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
	at java.lang.reflect.Method.invoke(Method.java:508)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:105)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:97)
	at java.util.concurrent.FutureTask.run(FutureTask.java:277)
	at java.lang.Thread.run(Thread.java:785)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-IbmJDK1.8/com.hazelcast$hazelcast-client/862/testReport/junit/com.hazelcast.client.quorum.cache/ClientCacheQuorumWriteTest/put_quorum_quorumType_WRITE_/

**ClientCacheQuorumWriteTest.putAll_noQuorum**
```
java.lang.Exception: Unexpected exception, expected<com.hazelcast.quorum.QuorumException> but was<com.hazelcast.cache.CacheNotExistsException>
	at org.junit.internal.runners.statements.ExpectException.evaluate(ExpectException.java:28)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:105)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:97)
	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)
	at java.util.concurrent.FutureTask.run(FutureTask.java:138)
	at java.lang.Thread.run(Thread.java:662)
Caused by: com.hazelcast.cache.CacheNotExistsException: Cache /hz/quorumaa358c09-393a-43bb-97fe-d6155275c56aWRITE is already destroyed or not created yet, on Member [127.0.0.1]:5805 - 5dbea9ca-8f02-4b7f-901b-02acb2b79ed0 this
	at com.hazelcast.cache.impl.AbstractCacheRecordStore.<init>(AbstractCacheRecordStore.java:131)
	at com.hazelcast.cache.impl.CacheRecordStore.<init>(CacheRecordStore.java:61)
	at com.hazelcast.cache.impl.CacheService.createNewRecordStore(CacheService.java:69)
	at com.hazelcast.cache.impl.CachePartitionSegment.createNew(CachePartitionSegment.java:52)
	at com.hazelcast.cache.impl.CachePartitionSegment.createNew(CachePartitionSegment.java:37)
	at com.hazelcast.util.ConcurrencyUtil.getOrPutSynchronized(ConcurrencyUtil.java:73)
	at com.hazelcast.cache.impl.CachePartitionSegment.getOrCreateRecordStore(CachePartitionSegment.java:74)
	at com.hazelcast.cache.impl.AbstractCacheService.getOrCreateRecordStore(AbstractCacheService.java:268)
	at com.hazelcast.cache.impl.operation.CachePutAllOperation.getServiceNamespace(CachePutAllOperation.java:143)
	at com.hazelcast.cache.impl.operation.CachePutAllOperation.getServiceNamespace(CachePutAllOperation.java:46)
	at com.hazelcast.quorum.impl.QuorumServiceImpl.getQuorumName(QuorumServiceImpl.java:191)
	at com.hazelcast.quorum.impl.QuorumServiceImpl.findQuorum(QuorumServiceImpl.java:181)
	at com.hazelcast.quorum.impl.QuorumServiceImpl.ensureQuorumPresent(QuorumServiceImpl.java:134)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.ensureQuorumPresent(OperationRunnerImpl.java:277)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:187)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:405)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:115)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:100)
	at ------ submitted from ------.(Unknown Source)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolve(InvocationFuture.java:127)
	at com.hazelcast.spi.impl.AbstractInvocationFuture$1.run(AbstractInvocationFuture.java:243)
	at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:895)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:918)
	at java.lang.Thread.run(Thread.java:662)
	at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:64)
	at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:80)
	at ------ submitted from ------.(Unknown Source)
	at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:96)
	at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:33)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.get(AbstractInvocationFuture.java:155)
	at com.hazelcast.client.cache.impl.AbstractClientCacheProxy.waitResponseFromAllPartitionsForPutAll(AbstractClientCacheProxy.java:355)
	at com.hazelcast.client.cache.impl.AbstractClientCacheProxy.putToAllPartitionsAndWaitForCompletion(AbstractClientCacheProxy.java:346)
	at com.hazelcast.client.cache.impl.AbstractClientCacheProxy.putAllInternal(AbstractClientCacheProxy.java:287)
	at com.hazelcast.client.cache.impl.AbstractClientCacheProxy.putAll(AbstractClientCacheProxy.java:278)
	at com.hazelcast.client.cache.impl.ClientCacheProxy.putAll(ClientCacheProxy.java:94)
	at com.hazelcast.client.cache.impl.ClientCacheProxy.putAll(ClientCacheProxy.java:201)
	at com.hazelcast.quorum.cache.CacheQuorumWriteTest.putAll_noQuorum(CacheQuorumWriteTest.java:152)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
	at java.lang.reflect.Method.invoke(Method.java:597)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.ExpectException.evaluate(ExpectException.java:19)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:105)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:97)
	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)
	at java.util.concurrent.FutureTask.run(FutureTask.java:138)
	at java.lang.Thread.run(Thread.java:662)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-freestyle/1612/testReport/junit/com.hazelcast.client.quorum.cache/ClientCacheQuorumWriteTest/putAll_noQuorum_quorumType_WRITE_/

**Comment by hasancelik** [comment]
```
java.util.concurrent.ExecutionException: com.hazelcast.cache.CacheNotExistsException: Cache /hz/quorumdccf8bee-ef47-4916-ac13-4e9caff312ffREAD_WRITE is already destroyed or not created yet, on Member [127.0.0.1]:9293 - a925ad08-95c0-4715-91f5-c35de888720a this
	at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:109)
	at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:33)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.get(AbstractInvocationFuture.java:183)
	at com.hazelcast.client.util.ClientDelegatingFuture.get(ClientDelegatingFuture.java:137)
	at com.hazelcast.client.util.ClientDelegatingFuture.get(ClientDelegatingFuture.java:128)
	at com.hazelcast.quorum.cache.CacheQuorumWriteTest.getAndRemoveAsync_quorum(CacheQuorumWriteTest.java:197)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:105)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:97)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.hazelcast.cache.CacheNotExistsException: Cache /hz/quorumdccf8bee-ef47-4916-ac13-4e9caff312ffREAD_WRITE is already destroyed or not created yet, on Member [127.0.0.1]:9293 - a925ad08-95c0-4715-91f5-c35de888720a this
	at com.hazelcast.client.impl.protocol.task.cache.AbstractCacheMessageTask.getOperationProvider(AbstractCacheMessageTask.java:51)
	at com.hazelcast.client.impl.protocol.task.cache.CacheGetAndRemoveMessageTask.prepareOperation(CacheGetAndRemoveMessageTask.java:45)
	at com.hazelcast.client.impl.protocol.task.AbstractPartitionMessageTask.processMessage(AbstractPartitionMessageTask.java:62)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.initializeAndProcessMessage(AbstractMessageTask.java:123)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.run(AbstractMessageTask.java:103)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:155)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:125)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:100)
	at ------ submitted from ------.(Unknown Source)
	at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:96)
	at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:33)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.get(AbstractInvocationFuture.java:183)
	at com.hazelcast.client.util.ClientDelegatingFuture.get(ClientDelegatingFuture.java:137)
	at com.hazelcast.client.util.ClientDelegatingFuture.get(ClientDelegatingFuture.java:128)
	at com.hazelcast.quorum.cache.CacheQuorumWriteTest.getAndRemoveAsync_quorum(CacheQuorumWriteTest.java:197)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:105)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:97)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-OracleJDK8/com.hazelcast$hazelcast-client/1448/testReport/junit/com.hazelcast.client.quorum.cache/ClientCacheQuorumWriteTest/getAndRemoveAsync_quorum_quorumType_READ_WRITE_/

**Comment by ahmetmircik** [comment]
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/13625/

**Comment by jerrinot** [comment]
some initial thinking: 
- there are various methods failing. they are all client-related. when the test is driven by a member proxies then it pass fine
- the failure is always `CacheNotExistsException`
- the client obtains a proxy via `instance.getCacheManager().getCache(CACHE_NAME + quorumType.name());` I wonder how eagierly this create the cache and distribute this information to all cluster members. Perhaps the cache is created on 1 member and this information is asynchronously distributed to remaining members? and if a client hits one of the remaining members before they learn about the new cache then it fails?

**Comment by Donnerbart** [comment]
Last failure 5 days ago, Vassilis fixed this issue 4 days ago: https://github.com/hazelcast/hazelcast/pull/12359

Could this be related? I think the issue caused some other test failures as well.

**Comment by jerrinot** [comment]
@Donnerbart: you were right. it was indeed fixed by the Vassilis PR. thanks for the hint!

**Comment by jerrinot** [comment]
fixed by #12359
